### PR TITLE
[MM-53102] Add feature keywords highlight without notification

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
- Adds a new section under settings/notifications for adding custom multi word keywords that gets highlighted without notification
- Adds a new classname for highlighting words although the styling is same as mentions highlights
- Added few components to ReduxFromProps pattern
- Adds supported type for hook of PluginComponent type
- Add upsel for highlight without notification
- Moved 'setting_item.tsx' to components folder
- Improved prop names and function structure for setting_item, setting_item_max and setting_item_min
- Moved 'toggle_modal_button.tsx' to components folder
- Removed t and utility messages from few components
- Fixed bug where tooltip was not getting rendered on restrictedButtons
- Improved the mobile view of the settings modal
- Adds E2E for the feature

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53102
https://mattermost.atlassian.net/browse/MM-53808
https://mattermost.atlassian.net/browse/MM-55145
https://mattermost.atlassian.net/browse/MM-53480

#### Screenshots

https://github.com/mattermost/mattermost/assets/17708702/08c2e3f5-f42b-42e3-b89b-649c65c97850

#### Release Note

```release-note
Added custom keywords with highlights without notification feature.
```

### CODE REVIEW

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts [hunk 1]
```diff
@@ -221,6 +221,16 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
...
 
+        // Add the only highlight without notification keyword
+        if (user.notify_props.highlight_keys) {
+            keys = keys.concat(user.notify_props.highlight_keys.split(',').map((key) => {
+                return {
+                    key,
+                    caseSensitive: true,
+                };
+            }));
+        }
```
Comment (marianunez) [inline webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:232]: This seems to work to highlight the words but it seems that this `getCurrentUserMentionKeys` is also called to list in the RHS actions when opening the Mentions RHS. I don't think we want to consider these highlights as mentions given they are not being notified about it

File: webapp/channels/src/components/markdown/index.ts [hunk 2]
```diff
@@ -11,27 +13,79 @@ import {getPost} from 'mattermost-redux/selectors/entities/posts';
...
 import Markdown from './markdown';
 
-type Props = {
+export type OwnProps = {
     channelNamesMap?: ChannelNamesMap;
     mentionKeys?: MentionKey[];
+    highlightKeys?: UserHighlightKey[];
     postId?: string;
+
+    /*
```
Comment (hmhealey) [inline webapp/channels/src/components/markdown/index.ts:33]: I'm liking the `PropsFromRedux` pattern here, but I feel like it'd be better to leave the rest of these props in the component file instead of moving them here. That does mean that we might have to repeat a few fields when a component has an `OwnProps` field, but I think that's fine since most components don't use `OwnProps`. I think `connect` is actually a bit less performant if you use `ownProps` because it has to rerun `mapStateToProps` even if `state` doesn't change then.

Could you move the rest of these fields back to the `Props` type in the component file?

File: webapp/channels/src/components/post_markdown/index.ts [hunk 3]
```diff
@@ -54,10 +54,30 @@ export function makeGetMentionKeysForPost(): (
     );
 }
 
-type OwnProps = {
-    channelId: string;
-    mentionKeys: MentionKey[];
+export type OwnProps = {
+
+    /**
```
Comment (hmhealey) [inline webapp/channels/src/components/post_markdown/index.ts:59]: Same here

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts [hunk 4]
```diff
@@ -223,6 +223,27 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
     },
 );
 
+export type UserHighlightKey = {
+    key: string;
+    caseSensitive?: boolean;
```
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:228]: Does having this field make using the existing code in `text_formatting.ts` easier, or can we leave it out? I can't imagine people will ever want this to be case sensitive, but there's one case that we need case sensitive mentions (just for first names and just because some people will be named words like Will that overlap with common words)

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts [hunk 5]
```diff
@@ -223,6 +223,27 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
...
+
+export const getUserHighlightKeys: (state: GlobalState) => UserHighlightKey[] = createSelector(
+    'getUserHighlightKeys',
+    getCurrentUser,
+    (user: UserProfile) => {
+        const highlightKeys: UserHighlightKey[] = [];
+
+        if (user.notify_props.highlight_keys.length > 0) {
+            user.notify_props.highlight_keys.split(',').forEach((key) => {
+                highlightKeys.push({key, caseSensitive: true});
```
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:239]: Speaking of that, I think we'd actually want these highlight keys to not be case sensitive

File: webapp/channels/src/components/markdown/index.ts (hunk 2 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/markdown/index.ts:33]: done moved to the component

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts (hunk 5 above)
Comment (johndavidlugtu) [reply webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:239]: +1 on keys not being case sensitive

cc: @abhijit-singh

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts (hunk 4 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts]: removed case sensitive key

**Review by hmhealey — APPROVED** [review]

**Review by marianunez — CHANGES_REQUESTED** [review]
Haven't done a full code review yet but it seems that the webapp is crashing when opening the user settings modal with 
```
 Uncaught TypeError: Cannot read properties of undefined (reading 'length')
    at St (user_settings_notifications.tsx:173:52
 ```

**Comment by abhijit-singh** [comment]
The screencap looks good but facing the same issue as Maria. Will check again once that is fixed.

**Review by abhijit-singh — APPROVED** [review]
Looks awesome, thanks @M-ZubairAhmed!

**Comment by yasserfaraazkhan** [comment]
Hi @M-ZubairAhmed , 
putting down a test in zephyr  https://github.com/mattermost/mattermost-test-management/pull/119/files

while trying out the changes observed a behaviour,

- if user has multiple phrases, that begin with  same words, not all phrases get highlighted 
in below screenshots, we can see only the first phrase was being highlighted.


Screenshot 2023-10-05 at 11 35 08 PM

Screenshot 2023-10-05 at 11 35 14 PM

File: webapp/channels/src/i18n/en.json [hunk 6]
```diff
@@ -5430,6 +5430,10 @@
   "user.settings.notifications.header": "Notifications",
   "user.settings.notifications.icon": "Notification Settings Icon",
   "user.settings.notifications.info": "Desktop notifications are available on Edge, Firefox, Safari, Chrome and Mattermost Desktop Apps.",
+  "user.settings.notifications.keywordsWithHighlight.extraInfo": "These keywords will be shown to you with a highlight when anyone sends a message that includes them.",
+  "user.settings.notifications.keywordsWithHighlight.inputTitle": "Enter non case-sensitive keywords, press Tab or use commas to separate them:",
+  "user.settings.notifications.keywordsWithHighlight.none": "None",
+  "user.settings.notifications.keywordsWithHighlight.title": "Keywords That Get Highlighted (without notifications)",
```
Comment (marianunez) [inline webapp/channels/src/i18n/en.json:5436]: @abhijit-singh seems there is inconsistency in casing with this title. Should the words within the parenthesis also be in capitals?

**Review by marianunez — CHANGES_REQUESTED** [review]
@M-ZubairAhmed overall code LGTM - however, requesting changes due to the ongoing conversation regarding moving this to a Professional feature

**Review by yasserfaraazkhan — CHANGES_REQUESTED** [review]
- when admin opens up `Start trial` or `purchase modal`, the notification modal stays on top.
- special character are not highlighted. (this is expected, right ?)
- IF a sentence ends with `DOT` or character, its not highlighted. ( is this expected ?)

image

https://github.com/mattermost/mattermost/assets/15979783/3b6f325d-fcab-4660-a8a2-f9af4835a006

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh 
> In the default state, there should not be a grey background on the section. The grey background should only be there on the hover state, just like for other items.

Changed


> Can we trigger the Professional feature tooltip when the user hovers anywhere on the entire section instead of just on the professional tag?

In the code the restricted button encapsulates all the logic for license check etc and contains the tooltip as well, so it would be hard to bring out the tooltip from that to the item. If this is not too much of a blocker can we skip this? It matches the current behavior of user groups in the global header menu item.


> In the default view, can we truncate the text that shows what has been set if it overflows, similar to how it happens for Keywords That Trigger Notification

Done

> Can we make the capitalization consistent for both the sections around keywords

Done

> I also noticed that email IDs also don't get highlighted in this new functionality

Added

> when admin opens up Start trial or purchase modal, the notification modal stays on top

Fixed

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx [hunk 7]
```diff
@@ -147,9 +150,10 @@ function getDefaultStateFromProps(props: Props): State {
     let channelKey = false;
     let isCustomKeysWithNotificationInputChecked = false;
     const customKeysWithNotification: MultiInputValue[] = [];
+    const customKeysWithHighlight: MultiInputValue[] = [];
 
     if (props.user.notify_props) {
-        if (props.user.notify_props.mention_keys) {
+        if (props.user.notify_props.mention_keys.length > 0) {
```
Comment (marianunez) [inline webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:156]: Should we add a null check here in case just in case? Before it was checking directly if it existed ? Also, why this change?

```suggestion
        if (props.user.notify_props.mention_keys?.length > 0) {
```

File: webapp/channels/src/utils/text_formatting.test.ts [hunk 8]
```diff
@@ -304,6 +306,101 @@ describe('highlightCurrentMentions', () => {
...
 });
 
+describe('highlightWithoutNotificationKeywords', () => {
+    test('should replace highlight keywords with tokens', () => {
+        const text = 'This is a test message with some keywords';
+        const tokens = new Map();
+        const highlightKeys = [
+            {key: 'keywords'},
+            {key: 'test'},
+        ];
```
Comment (marianunez) [inline webapp/channels/src/utils/text_formatting.test.ts:316]: Should we include phrases here with spaces as well for this test?

**Review by marianunez — APPROVED** [review]
Great work @M-ZubairAhmed! 🚀 Minor nits below

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 7 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:154]: changed

File: webapp/channels/src/utils/text_formatting.test.ts (hunk 8 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/utils/text_formatting.test.ts:316]: added a test

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh @yasserfaraazkhan gentle ping on this one

**Review by yasserfaraazkhan — COMMENTED** [review]
self-hosted:
- tested with uploading license. With sysadmin and user's account. On desktop and web
- tested with instance being on Enterprise trial
@johndavidlugtu @M-ZubairAhmed @marianunez 
If the Trial has ended, should the customer be able to use the highlighted keywords? 
It can also be a scenario that which the admin uploads a trial license and removes it. In between users have added keywords. These get highlighted.

image

**Review by abhijit-singh — APPROVED** [review]
Looks good to me, thanks for the great work and all the iterations on this @M-ZubairAhmed! 

To @yasserfaraazkhan's point - 
> If the Trial has ended, should the customer be able to use the highlighted keywords?

I believe we should stop highlighting keywords in this case but if possible maintain the keywords data added by the user. If that's complex to do, I'd vote for getting rid of the keywords altogether and disable the feature if they don't have the required license.

**Comment by M-ZubairAhmed** [comment]
@yasserfaraazkhan 

> If the license is E10 and E20, feature is enabled. But when saving the words. we see this error.

As per docs, both the licences are on the end of life https://docs.mattermost.com/about/editions-and-offerings.html#other-mattermost-plans

Added the fix for the highlight words not to get highlighted after the trial period.

**Comment by M-ZubairAhmed** [comment]
> Hi @M-ZubairAhmed the fixes work well now.
> 
> I was testing the admin notification part. It shows users requested and on which plan is available, but the feature name is missing. Please let me know this would be a follow up fix.
> 
> 

Thank you for pointing this one out. I thought the markdown came from the server but i was wrong this custom render is in webapp

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you for the changes @M-ZubairAhmed 
the notify admin part also works as expected.

image

**Comment by M-ZubairAhmed** [comment]
@johndavidlugtu can you please review

File: webapp/channels/src/components/user_settings/notifications/index.ts [hunk 9]
```diff
@@ -18,11 +22,27 @@ const mapStateToProps = (state: GlobalState) => {
...
+    const isStarterSKULicense = license.IsLicensed === 'true' && license.SelfHostedProducts === SelfHostedProducts.STARTER;
+
+    const isStarterFree = isCloudStarterFree || isSelfHostedStarter || isStarterSKULicense;
+
     return {
         sendPushNotifications,
         enableAutoResponder,
         isCollapsedThreadsEnabled: isCollapsedThreadsEnabled(state),
         isCallsRingingEnabled: isCallsEnabled(state, '0.17.0') && isCallsRingingEnabledOnServer(state),
+        isStarterFree,
```
Comment (hmhealey) [inline webapp/channels/src/components/user_settings/notifications/index.ts:44]: I guess my previous comments apply here too

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx [hunk 10]
```diff
@@ -145,9 +148,10 @@ function getDefaultStateFromProps(props: Props): State {
     let channelKey = false;
     let isCustomKeysWithNotificationInputChecked = false;
     const customKeysWithNotification: MultiInputValue[] = [];
+    const customKeysWithHighlight: MultiInputValue[] = [];
 
     if (props.user.notify_props) {
-        if (props.user.notify_props.mention_keys) {
+        if (props.user.notify_props?.mention_keys?.length > 0) {
```
Comment (hmhealey) [inline webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:154]: I don't think these `?.` are needed since we null checked `props.user.notify_props` above

File: webapp/channels/src/components/widgets/menu/menu_items/restricted_indicator.tsx [hunk 11]
```diff
@@ -107,27 +110,29 @@ const RestrictedIndicator = ({
...
-                            messageEndUser,
-                            customSecondaryButton: customSecondaryButtonInModal,
-                            feature,
-                            minimumPlanRequiredForFeature,
-                        }}
-                    >
-                        {icon}
-                        {ctaExtraContent}
-                    </ToggleModalButton>
+                    <span>
```
Comment (hmhealey) [inline webapp/channels/src/components/widgets/menu/menu_items/restricted_indicator.tsx:113]: Do we need this extra span?

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts [hunk 12]
```diff
@@ -224,6 +224,26 @@ export const getCurrentUserMentionKeys: (state: GlobalState) => UserMentionKey[]
...
+    key: string;
+}
+
+export const getHighlightWithoutNotificationKeys: (state: GlobalState) => HighlightWithoutNotificationKey[] = createSelector(
+    'getHighlightWithoutNotificationKeys',
+    getCurrentUser,
+    (user: UserProfile) => {
+        const highlightKeys: HighlightWithoutNotificationKey[] = [];
+
+        if (user?.notify_props?.highlight_keys?.length > 0) {
```
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:237]:
```suggestion
        if (user.notify_props.highlight_keys?.length > 0) {
```

File: webapp/channels/src/utils/text_formatting.tsx [hunk 13]
```diff
@@ -704,6 +715,79 @@ export function highlightCurrentMentions(
...
+        tokens.set(alias, {
+            value: `<span class="non-notification-highlight">${highlightKey}</span>`,
+            originalText: highlightKey,
+        });
+
+        return prefix + alias + suffix;
+    }
+
+    highlightKeys.
+        sort((a, b) => b.key.length - a.key.length).
```
Comment (hmhealey) [inline webapp/channels/src/utils/text_formatting.tsx:769]: I forget where we landed on fixing handling overlapping highlights since we discussed that last week. Did we decide this method of fixing it is enough for now?

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by johndavidlugtu** [comment]
Screenshot 2023-11-10 at 10 38 07 AM

[Admin view] Can we update the second sentence of the body to, "Upgrade to Professional plan to unlock this feature."


Screenshot 2023-11-10 at 10 43 14 AM
[End user view] Change second sentence of the body to, "Request your admin to upgrade to Mattermost Professional to unlock this feature."

@M-ZubairAhmed really great work here!

File: webapp/channels/src/components/post_markdown/post_markdown.test.tsx [hunk 14]
```diff
@@ -22,6 +24,14 @@ describe('components/PostMarkdown', () => {
...
         currentTeam: TestHelper.getTeamMock(),
         hideGuestTags: false,
+        isMilitaryTime: false,
+        timezone: '',
+        highlightKeys: [],
+        hasPluginTooltips: false,
+        isUserCanManageMembers: false,
+        isStarterFree: true,
+        isEnterpriseReady: false,
+        dispatch: jest.fn(),
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post_markdown/post_markdown.test.tsx:34]: Yes, I rechecked it's due to `makeMapStateToProps` along with PropsWithRedux.

File: webapp/channels/src/components/widgets/menu/menu_items/restricted_indicator.tsx (hunk 11 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/widgets/menu/menu_items/restricted_indicator.tsx:113]: yes it is https://github.com/mattermost/mattermost/pull/24905#discussion_r1360788142

File: webapp/channels/src/utils/text_formatting.tsx (hunk 13 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/utils/text_formatting.tsx:769]: yes will open a ticket for the case

File: webapp/channels/src/components/post_markdown/index.ts [hunk 15]
```diff
@@ -54,30 +56,46 @@ export function makeGetMentionKeysForPost(): (
...
             pluginHooks: state.plugins.components.MessageWillFormat,
             hasPluginTooltips: Boolean(state.plugins.components.LinkTooltip),
             isUserCanManageMembers: channel && canManageMembers(state, channel),
             mentionKeys: getMentionKeysForPost(state, ownProps.post, channel),
+            highlightKeys: getHighlightWithoutNotificationKeys(state),
             isMilitaryTime: getBool(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.USE_MILITARY_TIME, false),
             timezone: getCurrentTimezone(state),
             hideGuestTags: getConfig(state).HideGuestTags === 'true',
+            isStarterFree,
+            isEnterpriseReady,
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post_markdown/index.ts:92]: We need a distinction between an empty array and one with no license for scenarios.

**Comment by M-ZubairAhmed** [comment]
> Screenshot 2023-11-10 at 10 38 07 AM
> [Admin view] Can we update the second sentence of the body to, "Upgrade to Professional plan to unlock this feature."
> 
> @M-ZubairAhmed really great work here!
> 
> Pre-approving assuming the copy will be fixed, thanks

Changed

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 10 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:154]: Yeah once the app crashed on new user because the prop `highlight_keys` was never set for it, so i added these checks

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts (hunk 12 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/packages/mattermost-redux/src/selectors/entities/users.ts:237]: changed

File: webapp/channels/src/components/user_settings/notifications/index.ts (hunk 9 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/index.ts:44]: yes added selector

File: webapp/channels/src/components/post_markdown/index.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/post_markdown/index.ts]: I'm a bit surprised that this code would need it, but I can see that the setting screen would. It's totally fine as-is though

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 10 above)
Comment (hmhealey) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:154]: I think we needed to check for `highlight_keys`  specifically because it didn't previously exist in the `UserNotifyProps`, but I don't think we need to null check `notify_props` or `mention_keys`

File: webapp/channels/src/utils/license_utils.ts [hunk 16]
```diff
@@ -103,3 +104,14 @@ export const licenseSKUWithFirstLetterCapitalized = (license: ClientLicense) =>
     const sku = license.SkuShortName;
     return sku.charAt(0).toUpperCase() + sku.slice(1);
 };
+
+export function isEnterpriseOrCloudOrSKUStarterFree(license: ClientLicense, subscriptionProduct: Product | undefined, isEnterpriseReady: boolean) {
```
Comment (hmhealey) [inline webapp/channels/src/utils/license_utils.ts:108]: 0/5 but since the license, product, and `isEnterpriseReady` all come from redux, I think we could turn `isEnterpriseOrCloudOrSKUStarterFree` into a full selector that gets all of those from redux.

Also, I'm still trying to understand what we're trying to do by checking for these. Are we trying to determine if we can upsell the user on Enterprise?

**Review by hmhealey — APPROVED** [review]
One more comment on the code that checks the license, but it's nothing blocking, so this looks good to me

File: webapp/channels/src/utils/license_utils.ts (hunk 16 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/utils/license_utils.ts:108]: this checks if we are on a starter licence, if yes then we are checking if this is an enterprise ready app then we are adding an upsel

**Comment by M-ZubairAhmed** [comment]
> Successfully triggered E2E testing! [GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/387111) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/387111)

Only 1 failure not related to this change
